smart_card_connector_app/src/reader-tracker.js

```diff
@@ -0,0 +1,181 @@
+ * living in the NaCl space.
+ * @param {!goog.messaging.AbstractChannel} messageChannel
+ * @param {!goog.log.Logger} parentLogger
+ * @constructor
+ */
+GSC.ReaderTracker = function(messageChannel, parentLogger) {
+  /** @private */
+  this.logger_ = parentLogger;
+
+  messageChannel.registerService(
+      'reader_init_add', this.readerInitAddListener_.bind(this));
+  messageChannel.registerService(
+      'reader_finish_add', this.readerFinishAddListener_.bind(this));
+  messageChannel.registerService(
+      'reader_remove', this.readerRemoveListener_.bind(this));
+
+  /**
+   * @type {!goog.structs.Map.<string, !GSC.ReaderInfo>}
+   * @private
+   */
+  this.readers_ = new goog.structs.Map;
+
+  /**
+   * @type {!Array.<!function()>}
+   * @private
+   */
+  this.onUpdateListeners_ = [];
+
+  this.logger_.fine('Initialized');
+};
+
+/** @const */
+var ReaderTracker = GSC.ReaderTracker;
+
+/**
+ * TODO: I need to figure out the param because the closure compiler complains
+ *       about mismatched types, will do it at a later time.
+ * param {Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerInitAddListener_ = function(message) {
+  this.logger_.warning('readerInitAddListener_ ' + message);
+
+  // TODO: needs a promise? (check json.js)
+  var processor = new goog.json.NativeJsonProcessor;
+  var json = /** @type{Object} */ (processor.parse(message));
+  var name = json['readerName'];
+  var port = json['port'];
+  //var device = json['device'];
+
+  this.readers_.set(port, new ReaderInfo(name, 'yellow'));
+  this.fireOnUpdateListeners_();
+};
+
+/**
+ * param {Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerFinishAddListener_ = function(message) {
+  this.logger_.warning('readerFinishAddListener_ ' + message);
+
+  var processor = new goog.json.NativeJsonProcessor;
+  var json = /** @type{Object} */ (processor.parse(message));
+  var name = json['readerName'];
+  var port = json['port'];
+  var device = json['device'];
+  var returnCode = json['returnCode'];
+
+  // TODO: Make sure that this reader had already been seen in readerInitAddListener_
+  var value = new ReaderInfo(name, 'green');
+
+  if (returnCode !== 0) {
+    value.status = 'red';
+    // TODO: Send an actual error message, and not just the error id.
+    value.error = returnCode;
+  }
+
+  this.readers_.set(port, value);
+  this.fireOnUpdateListeners_();
+};
+
+/**
+ * param {Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerRemoveListener_ = function(message) {
+  this.logger_.warning('readerRemoveListener_ ' + message);
+
+  var processor = new goog.json.NativeJsonProcessor;
+  var json = /** @type{Object} */ (processor.parse(message));
+
+  var name = json['readerName'];
+  var port = json['port'];
+
+  this.readers_.remove(port);
+  this.fireOnUpdateListeners_();
+};
+
+/**
+ * param {function()} listener
+ */
+ReaderTracker.prototype.addOnUpdateListener = function(listener) {
```

Comment:
I think you also need to add a way to remove a listener.
Because, first, each time when the app window is opened, a new listener would be added without removing the old one. Second, all old listeners from the closed windows would be still called on each update, which sounds quite weird (won't they crash while trying to update the contents of the already destroyed window?).